src/dsl/sc.rs

```diff
@@ -1,57 +1,109 @@
-use std::hash::Hash;
+use std::{hash::Hash, rc::Rc};
 
 use halo2_proofs::arithmetic::Field;
 
 use crate::{
     ast::Circuit,
     compiler::{
-        cell_manager::CellManager, step_selector::StepSelectorBuilder, Compiler, CompilerConfig,
+        self,
+        cell_manager::{CellManager, SingleRowCellManager},
+        step_selector::{SimpleStepSelectorBuilder, StepSelectorBuilder},
+        CompilationUnit, Compiler, CompilerConfig,
     },
     ir::{
         assigments::AssigmentGenerator,
         sc::{MappingContext, SuperCircuit},
     },
 };
 
+use super::{lb::LookupTableRegistry, CircuitContext};
+
 pub struct SuperCircuitContext<F, MappingArgs> {
-    sc: SuperCircuit<F, MappingArgs>,
+    super_circuit: SuperCircuit<F, MappingArgs>,
+    sub_circuit_phase1: Vec<CompilationUnit<F>>,
+    tables: LookupTableRegistry<F>,
 }
 
 impl<F, MappingArgs> Default for SuperCircuitContext<F, MappingArgs> {
     fn default() -> Self {
         Self {
-            sc: Default::default(),
+            super_circuit: Default::default(),
+            sub_circuit_phase1: Default::default(),
+            tables: LookupTableRegistry::default(),
         }
     }
 }
 
 impl<F: Field + Hash, MappingArgs> SuperCircuitContext<F, MappingArgs> {
-    pub fn add<CM: CellManager, SSB: StepSelectorBuilder, TraceArgs>(
+    pub fn sub_circuit<CM: CellManager, SSB: StepSelectorBuilder, TraceArgs, Imports, Exports, D>(
         &mut self,
         config: CompilerConfig<CM, SSB>,
-        sub_circuit: Circuit<F, TraceArgs>,
-    ) -> AssigmentGenerator<F, TraceArgs> {
+        sub_circuit_def: D,
+        imports: Imports,
+    ) -> (AssigmentGenerator<F, TraceArgs>, Exports)
+    where
+        D: Fn(&mut CircuitContext<F, TraceArgs>, Imports) -> Exports,
+    {
+        let mut sub_circuit_context = CircuitContext {
+            circuit: Circuit::default(),
+            tables: self.tables.clone(),
+        };
+
+        let exports = sub_circuit_def(&mut sub_circuit_context, imports);
+
+        let sub_circuit = sub_circuit_context.circuit;
+
         let compiler = Compiler::from(config);
 
-        let (ir, assignment) = compiler.compile(&sub_circuit);
+        let (unit, assignment) = compiler.compile_phase1(&sub_circuit);
+        let assignment = assignment.unwrap_or_else(|| AssigmentGenerator::empty(unit.uuid));
 
-        self.sc.add_sub_circuit(ir);
+        self.sub_circuit_phase1.push(unit);
 
-        assignment.expect("circuit without trace")
+        (assignment, exports)
     }
 
     pub fn mapping<D: Fn(&mut MappingContext<F>, MappingArgs) + 'static>(&mut self, def: D) {
-        self.sc.set_mapping(def);
+        self.super_circuit.set_mapping(def);
+    }
+
+    fn compile(mut self) -> SuperCircuit<F, MappingArgs> {
+        // TODO: This is a hack, tbe compiler should be stateless
```

Comment:
Typo: "the" instead of "tbe"

Reply:
What do you mean "stateless" compiler here? There's only one compile for the super circuit, so it should share the same state throughout the compile?

Reply:
Fixed typo.

Compiler does not need to use &self in the future, because it does not make sense to keep state between compilations.

Reply:
Got it!